cpp/ParseFasm.cpp

```diff
@@ -0,0 +1,414 @@
+  // Tags: octal (o)
+  virtual Any visitOctalValue(FasmParser::OctalValueContext *context) override {
+    std::ostringstream data;
+    std::string value = context->OCTAL_VALUE()->getText();
+    auto it = value.begin();
+    it += 2; // skip 'b
+
+    // Build up a Num 3 bits at a time.
+    // Note that since the word size is not evenly divisible by 3,
+    // intermediate values can be greater than the word size.
+    // This is why the 'word' below is 64 bits wide.
+    int bits = lead_bits(count_without(it, value.end(), '_') * 3);
+    uint64_t word = 0; // could temporarily overflow uint32_t
+    while (it != value.end()) {
+      if (*it != '_') {
+        word = (word << 3) | (*it - '0');
+        bits += 3;
+        if (bits >= Num::width) {
+          data << Num(word >> (bits - Num::width));
+          word >>= Num::width;
+          bits -= Num::width;
+        }
+      }
+      it++;
+    }
+    assert(!word);
+    return withHeader('o', data.str());
+  }
+
+  // A collection of annotations. { ... }
+  // Tags: annotations ({)
+  virtual Any visitAnnotations(FasmParser::AnnotationsContext *context) override {
+    std::ostringstream data;
+    for (auto &a : context->annotation()) {
+      data << visit(a).as<std::string>();
+    }
+    return withHeader('{', data.str());
+  }
+
+  // An annotation: x = "y"
+  // Tags: annotation (a), annotation name (.), annotation value (=)
+  virtual Any visitAnnotation(FasmParser::AnnotationContext *context) override {
+    std::ostringstream data;
+    data << Str('.', context->ANNOTATION_NAME()->getText());
+    if (context->ANNOTATION_VALUE()) {
+      std::string value = context->ANNOTATION_VALUE()->getText();
+      value.erase(0, 1);  // Convert "value" -> value
+      value.pop_back();
+      data << Str('=', value);
+    }
+    return withHeader('a', data.str());
+  }
+
+private:
+
+  std::ostream &out;
+};
+
+
+// Common portion of 'from_string' and 'from_file'.
+// Consumes an input stream and produces an output stream.
+static
+void parse_fasm(std::istream &in, std::ostream &out) {
+  ANTLRInputStream stream(in);
+  FasmLexer lexer(&stream);
+  CommonTokenStream tokens(&lexer);
+  FasmParser parser(&tokens);
+  auto *tree = parser.fasmFile();
+  FasmParserBaseVisitor(out).visit(tree);
+}
+
+// This static string is used to temporarily hold the result
+// returned from external functions.
+// This is to allow Python consumers, because I could not
+// find a convenient way for Python to free a malloc'ed string.
+static std::string result;
+
+// Parse the given input string, returning output.
+// Use hex mode (see above) if hex is true.
+const char *from_string(const char *in, bool hex) {
+  hex_mode = hex;
+  std::istringstream input(in);
+  std::ostringstream output;
+  parse_fasm(input, output);
+  output.put(0);
+  result = output.str();
+  return result.c_str();
+}
+
+// Parse the given input file, returning output.
+// Use hex mode (see above) if hex is true.
```

Comment:
I think it would be really useful to create a top level block comment that explains how the string serialization is working here.  I think the approach is sound, but right now I have to read a lot of the code to follow along.  As a concrete suggestion, I think a couple examples of FASM in -> output stream might go a long way.

Reply:
Done.